### ISSUE TITLE
loading huge EXR image crashes

### ISSUE THREAD

**Issue by claudeha** [darktable-org/darktable#3298]
**Describe the bug**
`darktable 32000x18000.exr` → segmentation fault, and `darktable` can no longer be started from that directory without crashing

**To Reproduce**
1. Create an EXR file with very large image size (I tried 32000x18000 which crashes; 16000x9000 loads ok)
2. Try to load it in `darktable`
3. Crash

**Expected behavior**
Ideally, for the image to be loaded successfully.
Pragmatically, an error message saying that the dimensions are too big, without a crash.

**Platform (please complete the following information):**
 - OS: Debian Buster
 - Version: git master at `release-3.0.0rc0-22-gc03143cb9`

**Additional context**
My EXR files have R G B half channels.  They contain computer generated fractal imagery generated with https://mathr.co.uk/kf/kf.html whose output tiles were assembled into a large image using https://mathr.co.uk/exrtact

**Crash log**
[darktable_bt_72GVA0.txt](https://github.com/darktable-org/darktable/files/3811088/darktable_bt_72GVA0.txt)

**Comment by claudeha** [comment]
master branch at release-3.1.0-426-g88bf11bd2 still crashes
[darktable_bt_YULOE0.txt](https://github.com/darktable-org/darktable/files/4106393/darktable_bt_YULOE0.txt)

**Comment by johnny-bit** [comment]
Since #4210 is merged in master can you re-check?

**Comment by claudeha** [comment]
Still crashes with master branch at release-3.5.0-139-gca5a21527.

TL;DR 32bit signed int overflow at `darktable/src/common/imageio_exr.cc:132` gives a very huge size in `memset(buf, 0, 4 * img->width * img->height * sizeof(float));`  ($8 is the size it should be)

```
Thread 1 "darktable" received signal SIGSEGV, Segmentation fault.
__memset_avx2_unaligned_erms () at ../sysdeps/x86_64/multiarch/memset-vec-unaligned-erms.S:202
202	../sysdeps/x86_64/multiarch/memset-vec-unaligned-erms.S: No such file or directory.
(gdb) bt
#0  __memset_avx2_unaligned_erms () at ../sysdeps/x86_64/multiarch/memset-vec-unaligned-erms.S:202
#1  0x00007ffff7e49cf4 in dt_imageio_open_exr(dt_image_t*, char const*, dt_mipmap_buffer_t*)
    (img=img@entry=0x7fffffff6710, filename=filename@entry=0x7fffffff6c70 "/home/claude/tmp/drt/32000x18000.exr", mbuf=mbuf@entry=0x7fffffff7cc0)
    at /home/claude/code/github.com/darktable-org/darktable/src/common/imageio_exr.cc:132
#2  0x00007ffff7c737a8 in dt_imageio_open_hdr (buf=0x7fffffff7cc0, filename=0x7fffffff6c70 "/home/claude/tmp/drt/32000x18000.exr", img=0x7fffffff6710)
    at /home/claude/code/github.com/darktable-org/darktable/src/common/imageio.c:396
#3  dt_imageio_open_hdr (img=0x7fffffff6710, filename=0x7fffffff6c70 "/home/claude/tmp/drt/32000x18000.exr", buf=0x7fffffff7cc0) at /home/claude/code/github.com/darktable-org/darktable/src/common/imageio.c:384
#4  0x00007ffff7c7559e in dt_imageio_open (img=img@entry=0x7fffffff6710, filename=filename@entry=0x7fffffff6c70 "/home/claude/tmp/drt/32000x18000.exr", buf=buf@entry=0x7fffffff7cc0)
    at /home/claude/code/github.com/darktable-org/darktable/src/common/imageio.c:1178
#5  0x00007ffff7c9c3a7 in dt_mipmap_cache_get_with_caller
    (cache=0x7fffffff6710, buf=0x7fffffff7cc0, imgid=324, mip=DT_MIPMAP_FULL, flags=<optimized out>, mode=<optimized out>, file=0x7ffff7ed7730 "/home/claude/code/github.com/darktable-org/darktable/src/common/darktable.c", line=275) at /home/claude/code/github.com/darktable-org/darktable/src/common/mipmap_cache.c:784
#6  0x00007ffff7c26cdb in dt_load_from_string (input=0x144 <error: Cannot access memory at address 0x144>, open_image_in_dr=0, single_image=0x7fffffff802c)
    at /home/claude/code/github.com/darktable-org/darktable/src/common/darktable.c:275
#7  0x00007ffff7c2926b in dt_init (argc=<optimized out>, argv=<optimized out>, init_gui=init_gui@entry=1, load_data=load_data@entry=1, L=L@entry=0x0)
    at /home/claude/code/github.com/darktable-org/darktable/src/common/darktable.c:1183
#8  0x0000555555555086 in main (argc=<optimized out>, argv=<optimized out>) at /home/claude/code/github.com/darktable-org/darktable/src/main.c:92
(gdb) up
#1  0x00007ffff7e49cf4 in dt_imageio_open_exr (img=img@entry=0x7fffffff6710, filename=filename@entry=0x7fffffff6c70 "/home/claude/tmp/drt/32000x18000.exr", mbuf=mbuf@entry=0x7fffffff7cc0)
    at /home/claude/code/github.com/darktable-org/darktable/src/common/imageio_exr.cc:132
132	  memset(buf, 0, 4 * img->width * img->height * sizeof(float));
(gdb) print buf
$1 = (float *) 0x7ffcf9644080
(gdb) print img->width
$2 = 32000
(gdb) print img->height
$3 = 18000
(gdb) print sizeof(float)
$4 = 4
(gdb) print 4 * 32000 * 18000 * 4
$5 = 626065408
(gdb) print 4 * 32000 * 18000 * 4LL
$6 = -7963869184
(gdb) print 4 * 32000 * 18000 * 4ULL
$7 = 18446744065745682432
(gdb) print 4ULL * 32000 * 18000 * 4
$8 = 9216000000
(gdb) 
```

**Comment by claudeha** [comment]
this command might find other occurrences to audit, if darktable is intended support images larger than ~134Mpixel.

```
git grep -i "width[^,]*\*[^,]*height" |
grep -v "tools/regression_tests/src/web/assets/bootstrap/css/bootstrap.css.map:"
```

(also try swapping width and height in the regex, the `,` rejection avoids seeing lots of function declarations; the bootstrap.css.map is a huge flood of irrelevance)

**Comment by johnny-bit** [comment]
Hmmm...

that sounds plausible...

OK since i don't have any image larger that 134MPix how to create one using your linked tool i wonder if sane and simple approach here would be to change

```cpp
memset(buf, 0, 4 * img->width * img->height * sizeof(float));
```
to

```cpp
const size_t alloc_size = 4 * img->width * img->height * sizeof(float)
memset(buf, 0, alloc_size);
```
or
```cpp
memset(buf, 0, (size_t)4 * img->width * img->height * sizeof(float));
```

I think I prefer the size_t cast one ;)

can you check if either change "works"?

**Comment by johnny-bit** [comment]
OK, actually i could just compile ;)

and 3 versions:
```cpp
  const size_t t1 = (size_t)4 * 32000 * 18000 * sizeof(float);
  const size_t t2 = 4 * 32000 * 18000 * sizeof(float);
  const size_t t3 = sizeof(float) * 4 * 32000 * 18000;
```

t2 line fails with:

```
test.c:15:31: error: integer overflow in expression of type ‘int’ results in ‘-1990967296’ [-Werror=overflow]
   15 |   const size_t t2 = 4 * 32000 * 18000 * sizeof(float);
```

OK, I'm doing a PR... I still think I like t1 ver better (cast) so I'll go with it untill someone yells ;)

**Comment by johnny-bit** [comment]
Can you check PR #7419? Should fix some issues, but you may detect more as it's possible that there are places in code that would have incorrect allocation size calculation order.

**Comment by claudeha** [comment]
Here is a short C++ program using OpenEXR that can generate big images:

```
/*
g++ -o big-exr big-exr.cc `pkg-config --cflags --libs OpenEXR`
./big-exr 32000 18000 big-exr.exr
*/

#include <ImfNamespace.h>
#include <ImfOutputFile.h>
#include <ImfHeader.h>
#include <ImfChannelList.h>
#include <ImfArray.h>
#include <ImfFrameBuffer.h>
#include <cstdio>
#include <cstdlib>

namespace IMF = OPENEXR_IMF_NAMESPACE;
using namespace IMF;

int main(int argc, char **argv)
{
  if (argc != 4)
  {
    std::fprintf(stderr, "usage: %s width height out.exr\n", argv[0]);
    return 1;
  }
  const int width = atoi(argv[1]);
  const int height = atoi(argv[2]);
  const char *filename = argv[3];
  float *rgb = (float *) std::calloc(1, (size_t)3 * width * height * sizeof(float));
  if (! rgb)
  {
    return 1;
  }
  Header header(width, height);
  header.channels().insert("R", Channel(IMF::FLOAT));
  header.channels().insert("G", Channel(IMF::FLOAT));
  header.channels().insert("B", Channel(IMF::FLOAT));
  OutputFile of(filename, header);
  FrameBuffer fb;
  fb.insert("R", Slice(IMF::FLOAT, (char *) (rgb + 0), sizeof(*rgb) * 3, sizeof(*rgb) * 3 * width));
  fb.insert("G", Slice(IMF::FLOAT, (char *) (rgb + 1), sizeof(*rgb) * 3, sizeof(*rgb) * 3 * width));
  fb.insert("B", Slice(IMF::FLOAT, (char *) (rgb + 2), sizeof(*rgb) * 3, sizeof(*rgb) * 3 * width));
  of.setFrameBuffer(fb);
  of.writePixels(height);
  std::free(rgb);
  return 0;
}
```

**Comment by claudeha** [comment]
PR #7419 doesn't crash on load.  But trying to export it to 8bit PNG at 32000x18000 gets darktable Killed by Out Of Memory (it was using 9.3GB before trying to export, I have 32GB RAM) so I suspect another uncaught overflow somewhere,

**Comment by johnny-bit** [comment]
I've pushed couple more changes there... Please check now but I already see places where it still might fail :/

**Comment by claudeha** [comment]
PNG export still OOMs

**Comment by johnny-bit** [comment]
OK, so I went ahead and tried to hunt down whatever I could find...

I think I got ALL places. The issue though might be that your system OOMs because there can be bufor allocation per thread (so eg some option can allocate `sizeof(float) * 4 * 32000 * 18000 * numthreads` and have that aligned to 64 bytes so in 8-thread system than can grow to over 68+GBs... and as a workaround you can limit number of threads to 1 so see if dt OOMs on number of threads.

Try #7419 and if it still ooms try with single thread.
